lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs

```diff
@@ -1375,6 +1492,205 @@ makeChangeForCoin
+  -> NonEmpty TokenMap
+  -> NonEmpty TokenMap
+addMintValuesToChangeMaps mints =
+    F.foldr
+        (\mint -> (addMintValueToChangeMaps mint .))
+        id
+        (TokenMap.toFlatList mints)
+
+-- | Remove a burnt value from the change map. This function presumes that
+-- the change map is partially-ordered in ascending order (i.e. the largest
+-- element is at the end of the list). This function operates under the
+-- precondition that a valid selection has been performed, and so assumes
+-- that there is enough asset quantity in the change map to satisfy the
+-- burn. If there is not, it will burn as much as it can.
+--
+-- Properties:
+-- 1. wrt to the total value of assets in the changeMaps, calling this
+-- function is the same as removing the burnt value from the change maps:
+--   F.fold changeMaps `TokenMap.difference` TokenMap.singleton assetId qty
+--          == F.fold (f (assetId, qty) changeMaps)
+-- 2. length is conserved
+-- 3. order is preserved (i.e., the result is in ascending partial order).
+removeBurnValueFromChangeMaps
+    :: (AssetId, TokenQuantity)
+    -> NonEmpty TokenMap
+    -> NonEmpty TokenMap
+removeBurnValueFromChangeMaps (assetId, assetQty) =
+    burnWithGas assetQty
+  where
+      -- Subtract two token quantities, clamping negative values to zero.
+      subtract' :: TokenQuantity -> TokenQuantity -> TokenQuantity
+      subtract' x y =
+          fromMaybe TokenQuantity.zero (x `TokenQuantity.subtract` y)
+
+      burnWithGas
+          :: TokenQuantity
+          -> NonEmpty TokenMap
+          -> NonEmpty TokenMap
+      -- No need to burn anything
+      burnWithGas (TokenQuantity 0) = id
+      -- Must burn some still
+      burnWithGas gas =
+          let
+              toBurn = TokenMap.singleton assetId gas
+          in \case
+              -- We only have one element left to burn, try and burn the
+              -- entire gas qty from it
+              x :| [] ->
+                  (x `TokenMap.difference` toBurn) :| []
+              -- We have a few entries left to burn, burn what we can then
+              -- possibly continue burning
+              x :| (y : rest) ->
+                  let
+                      xs :: NonEmpty TokenMap
+                      xs = y :| rest
+
+                      burnt :: TokenMap
+                      burnt = x `TokenMap.difference` toBurn
+
+                      burnMore :: NonEmpty TokenMap
+                      burnMore = case (x `TokenMap.getQuantity` assetId) of
+                          -- Not enough, continue burning
+                          qty | qty < gas ->
+                              burnWithGas (gas `subtract'` qty) xs
+                          -- Burnt all the gas, don't burn any more
+                          _qty ->
+                              xs
+                  in
+                      burnt `NE.cons` burnMore
+
+removeBurnValuesFromChangeMaps
+  :: TokenMap
+  -> NonEmpty TokenMap
+  -> NonEmpty TokenMap
```

Comment:
Perhaps add a comment to this function, and comments for each of the arguments?

(Though the name pretty obviously indicates that it's a pluralization of `removeBurnValueFromChangeMaps`.)
```suggestion
    :: TokenMap
    -> NonEmpty TokenMap
    -> NonEmpty TokenMap
```

Reply:
Addressed in df9d84252